ledger/acctupdates.go

```
@@ -376,60 +523,537 @@ func (au *accountUpdates) getAssetCreatorForRound(rnd basics.Round, aidx basics.
+	rootHash, err := trie.RootHash()
+	if err != nil {
+		return rnd, fmt.Errorf("accountsInitialize was unable to retrieve trie root hash: %v", err)
+	}
+	if rootHash.IsZero() {
+		accountIdx := 0
+		for {
+			bal, err := encodedAccountsRange(tx, accountIdx, trieRebuildAccountChunkSize)
+			if err != nil {
+				return rnd, err
+			}
+			if len(bal) == 0 {
+				break
+			}
+			for _, balance := range bal {
+				addrBytes, accountBytes := balance.Address, balance.AccountData
+				var addr basics.Address
+				copy(addr[:], addrBytes)
+				var accountData basics.AccountData
+				err = protocol.Decode(accountBytes, &accountData)
+				if err != nil {
+					return rnd, err
+				}
+				hash := accountHashBuilder(addr, accountData, accountBytes)
+				added, err := trie.Add(hash)
+				if err != nil {
+					return rnd, fmt.Errorf("accountsInitialize was unable to add changes to trie: %v", err)
+				}
+				if !added {
+					au.log.Warnf("attempted to add duplicate hash '%v' to merkle trie.", hash)
+				}
+			}
+
+			// this trie Evict will commit using the current transaction.
+			// if anything goes wrong, it will still get rolled back.
+			_, err = trie.Evict(true)
+			if err != nil {
+				return 0, fmt.Errorf("accountsInitialize was unable to commit changes to trie: %v", err)
+			}
+			if len(bal) < trieRebuildAccountChunkSize {
+				break
+			}
+			accountIdx += trieRebuildAccountChunkSize
+		}
+	}
+	au.balancesTrie = trie
+	return rnd, nil
+}
+
+func (au *accountUpdates) accountsUpdateBalaces(accountsDeltas map[basics.Address]accountDelta) (err error) {
+	if au.catchpointInterval == 0 {
+		return nil
+	}
+	var added, deleted bool
+	for addr, delta := range accountsDeltas {
+		if !delta.old.IsZero() {
+			deleteHash := accountHashBuilder(addr, delta.old, protocol.Encode(&delta.old))
+			deleted, err = au.balancesTrie.Delete(deleteHash)
+			if err != nil {
+				return err
+			}
+			if !deleted {
+				au.log.Warnf("failed to delete hash '%v' from merkle trie", deleteHash)
```

here and in `if !added`, it seems like you should return an error, since it indicates that your data structure is corrupted.

---

I was thinking about it. I'm not sure if we want to.

---

I'm closing this conversation, but please do let me know if you think it's an issue.